### PR TITLE
Add CatBoostRanker & pool.get_group_id_hash() for ranking

### LINKED ISSUE #766: CatBoostRanker class in Python
XGBoost and LightGBM have sklearn interfaces for ranking, we should also introduce one.

xgboost.XGBRanker
lightgbm.LGBMRanker

### PR DESCRIPTION
**Pull request by AnnaAraslanova**
This Pull Request [closes #766](https://github.com/catboost/catboost/issues/766)

I hereby agree to the terms of the CLA available at: [link](https://yandex.ru/legal/cla/?lang=ru)

### CODE REVIEW

File: catboost/python-package/catboost/core.py [hunk 1]
```diff
@@ -5226,6 +5236,400 @@ def _get_default_prediction_type(self):
...
+                raise CatBoostError("Wrong initializing group_id: X is catboost.Pool object, group_id must be initialized inside catboost.Pool.")
+            group_id = X.get_group_id_hash()
+
+        if y is None:
+            raise CatBoostError("y must be initialized.")
+
+        predictions = self.predict(X)
+        return self._compute_mean_ndcg(predictions, y, group_id, metric_type, top, denominator_type)
+
+    def _compute_mean_ndcg(self, y_pred, y_true, group_id=None, metric_type='Base', top=int(pow(2, 32) - 1), denominator_type='LogPosition'):
```
Comment (nikitxskv) [inline catboost/python-package/catboost/core.py:5606]: Давай вместо того, что бы писать свою функцию будем использовать готовую:
https://github.com/catboost/catboost/blob/master/catboost/python-package/catboost/utils.py#L247
Там третьим параметром нужно передать имя метрики как строку. По умолчанию, если пользователь не передал никакие аргументы будем вызывать с metric='NDCG', если передал - будем добавлять к строке параметры metric='NDCG:top=5;type=Base',

File: catboost/python-package/catboost/core.py [hunk 2]
```diff
@@ -5226,6 +5236,400 @@ def _get_default_prediction_type(self):
...
+                value = _ComputeNDCG(y_pred[idxs], y_true[idxs], metric_type, top, denominator_type)
+                sum_ndcg += value
+
+            return sum_ndcg / len(queries)
+
+    def _check_is_ranking_loss(self, loss_function):
+        is_ranking = self._is_groupwise_objective(loss_function) or self._is_pairwise_objective(loss_function)
+        is_regression = self._is_regression_objective(loss_function) or self._is_multiregression_objective(loss_function)
+
+        if isinstance(loss_function, str) and is_regression:
```
Comment (nikitxskv) [inline catboost/python-package/catboost/core.py:5624]: Зачем тут и ниже проверки на тип, если они и так в is_ranking и is_regression зашиты?

File: catboost/python-package/catboost/core.py [hunk 3]
```diff
@@ -5226,6 +5236,400 @@ def _get_default_prediction_type(self):
...
+            for query in queries:
+                idxs = (group_id == query)
+                value = _ComputeNDCG(y_pred[idxs], y_true[idxs], metric_type, top, denominator_type)
+                sum_ndcg += value
+
+            return sum_ndcg / len(queries)
+
+    def _check_is_ranking_loss(self, loss_function):
+        is_ranking = self._is_groupwise_objective(loss_function) or self._is_pairwise_objective(loss_function)
+        is_regression = self._is_regression_objective(loss_function) or self._is_multiregression_objective(loss_function)
```
Comment (nikitxskv) [inline catboost/python-package/catboost/core.py:5622]: А на мультирегрессоре у нас ничего не сломается? Давай уберем на всякий случай? Кажется со score будут проблемы

File: catboost/python-package/catboost/core.py [hunk 4]
```diff
@@ -5226,6 +5236,400 @@ def _get_default_prediction_type(self):
...
+        verbose : bool
+            If True, writes the evaluation metric measured set to stderr.
+        Returns
+        -------
+        prediction : generator for each iteration that generates:
+            If data is for a single object, the return value is single float formula return value
+            otherwise one-dimensional numpy.ndarray of formula return values for each object.
+        """
+
+        if not isinstance(X, Pool):
```
Comment (nikitxskv) [inline catboost/python-package/catboost/core.py:5564]: В предсказани же group_id никак не участвует, так что кажется этот метод можно оставить дефолтным

File: catboost/python-package/catboost/core.py [hunk 5]
```diff
@@ -5226,6 +5236,400 @@ def _get_default_prediction_type(self):
...
+            If -1, then the number of threads is set to the number of CPU cores.
+        verbose : bool
+            If True, writes the evaluation metric measured set to stderr.
+        Returns
+        -------
+        prediction :
+            If data is for a single object, the return value is single float formula return value
+            otherwise one-dimensional numpy.ndarray of formula return values for each object.
+        """
+        if not isinstance(X, Pool):
```
Comment (nikitxskv) [inline catboost/python-package/catboost/core.py:5526]: В предсказани же group_id никак не участвует, так что кажется этот метод можно оставить дефолтным

File: catboost/python-package/catboost/core.py [hunk 6]
```diff
@@ -5226,6 +5236,400 @@ def _get_default_prediction_type(self):
...
+        _process_synonyms(params)
+        if 'loss_function' in params:
+            self._check_is_ranking_loss(params['loss_function'])
+
+        if ('loss_function' in params and
+            'max_depth' in params and
+             params['loss_function'] == 'PairLogitPairwise' and
+             params['max_depth'] >= 8
+           ):
+            raise Exception('max_depth for pair-logit-pairwise should be < 8')
```
Comment (nikitxskv) [inline catboost/python-package/catboost/core.py:5485]: А откуда это ограничение? Катбуст падает, если передать такие параметры? Если нет, то давай уберем это.

File: catboost/python-package/catboost/core.py [hunk 7]
```diff
@@ -5226,6 +5236,400 @@ def _get_default_prediction_type(self):
...
+
+        if not isinstance(X, Pool) and group_id is None:
+            group_id = np.zeros(X.shape[0], dtype=int)
```
Comment (nikitxskv) [inline catboost/python-package/catboost/core.py:5488]: Это тоже можно убрать, так как обучаться без групп странно и мы поидее должны падать, если группы не передали.

File: catboost/python-package/catboost/_catboost.pyx [hunk 8]
```diff
@@ -3952,6 +3952,16 @@ cdef class _PoolBase:
             return [weight for weight in non_trivial_data]
 
 
+    cpdef get_group_id_hash(self):
```
[the remark this answers is not in the document]
Comment (AnnaAraslanova) [reply catboost/python-package/catboost/_catboost.pyx:3955]: * теперь функция возвращает None, если группы не были переданы
 * _hash я не хочу удалять из названия, т.к. лучше явно указать, что там возвращаются хеши, а не переданные номера групп.  если убрать _hash, то пользователи будут думать, что get_group_id возвращает те же самые названия групп

File: catboost/python-package/catboost/core.py [hunk 9]
```diff
@@ -5226,6 +5236,381 @@ def _get_default_prediction_type(self):
...
+        if 'loss_function' in params:
+            self._check_is_ranking_loss(params['loss_function'])
+
+        self._fit(X, y, cat_features, text_features, embedding_features, pairs,
+                  sample_weight, group_id, group_weight, subgroup_id, pairs_weight, baseline, use_best_model,
+                  eval_set, verbose, logging_level, plot, column_description, verbose_eval, metric_period,
+                  silent, early_stopping_rounds, save_snapshot, snapshot_file, snapshot_interval, init_model)
+        return self
+
+    def predict(self, X, group_id=None, ntree_start=0, ntree_end=0, thread_count=-1, verbose=None):
```
[the remark this answers is not in the document]
Comment (AnnaAraslanova) [reply catboost/python-package/catboost/core.py:5480]: да, точно. я почему-то была уверена, что используются

File: catboost/python-package/catboost/core.py [hunk 10]
```diff
@@ -5226,6 +5236,381 @@ def _get_default_prediction_type(self):
...
+            raise CatBoostError("y must be initialized.")
+        group_id = group_id if group_id is not None else np.zeros(X.shape[0], dtype=int)
+
+        predictions = self.predict(X)
+        params = get_params_line([top, type, denominator], ['top', 'type', 'denominator'])
+        return _eval_metric_util([y], [predictions], "NDCG{}".format(params), None, group_id, group_weight, None, None, thread_count)[0]
+
+
+    def _check_is_ranking_loss(self, loss_function):
+        is_ranking = self._is_groupwise_objective(loss_function) or self._is_pairwise_objective(loss_function)
```
[the remark this answers is not in the document]
Comment (AnnaAraslanova) [reply catboost/python-package/catboost/core.py:5602]: справедливо

File: catboost/python-package/catboost/core.py [hunk 11]
```diff
@@ -1519,6 +1524,15 @@ def _is_regression_objective(self, loss_function):
     def _is_multiregression_objective(self, loss_function):
         return isinstance(loss_function, str) and is_multiregression_objective(loss_function)
 
+    def _is_groupwise_objective(self, loss_function):
+        return isinstance(loss_function, str) and is_groupwise_metric(loss_function)
+
+    def _is_pairwise_objective(self, loss_function):
+        return isinstance(loss_function, str) and is_pairwise_metric(loss_function)
+
```
Comment (nikitxskv) [inline catboost/python-package/catboost/core.py:1532]: Это же теперь не нужно? Давай удалим

File: catboost/python-package/catboost/core.py [hunk 12]
```diff
@@ -57,6 +57,8 @@ class Series(object):
 is_regression_objective = _catboost.is_regression_objective
 is_multiregression_objective = _catboost.is_multiregression_objective
 is_groupwise_metric = _catboost.is_groupwise_metric
+is_pairwise_metric = _catboost.is_pairwise_metric
```
Comment (nikitxskv) [inline catboost/python-package/catboost/core.py:60]: Это же теперь не нужно? Давай удалим

File: catboost/python-package/catboost/core.py [hunk 13]
```diff
@@ -5226,6 +5240,368 @@ def _get_default_prediction_type(self):
...
+            If this parameter is a string or pathlib.Path, load initial model from the path specified by this string.
+        Returns
+        -------
+        model : CatBoost
+        """
+
+        params = deepcopy(self._init_params)
+        _process_synonyms(params)
+        if 'loss_function' in params:
+            self._check_is_ranking_loss(params['loss_function'])
```
Comment (nikitxskv) [inline catboost/python-package/catboost/core.py:5476]: Слушай, а это ведь можно в конструктор перенести?
Давай так и сделаем, и у других классов тоже. Непонятно вообще, почему это тут.

File: catboost/python-package/catboost/core.py [hunk 14]
```diff
@@ -5825,3 +6201,21 @@ def to_classifier(model):
     for attr in model.__dict__:
         setattr(classifier, attr, getattr(model, attr))
     return classifier
+
+def to_ranker(model):
```
Comment (nikitxskv) [inline catboost/python-package/catboost/core.py:6205]: А давай подсократим немного кода тут.
У нас есть сейчас три почти одинаковые функции: to_ranker, to_classifier, to_regressor
Давай их внутренности вынесем в отдельную функцию _to_subclass(model, SubClass, check_objective, loss_function)

```
def to_ranker(model):
    return _to_subclass(model, CatBoostRanker, _check_is_ranking_loss, params.get('loss_function'))
```
и тд
При этом _check_is_ranking_loss и остальные функции можно вынести из _CatBoostBase,так как они self вообще никак не используют.

File: catboost/python-package/catboost/core.py (hunk 13 above)
Comment (AnnaAraslanova) [reply catboost/python-package/catboost/core.py:5476]: обсудили, что не надо

File: catboost/python-package/catboost/core.py (hunk 14 above)
Comment (AnnaAraslanova) [reply catboost/python-package/catboost/core.py:6205]: * функции проверки лоссов сделала static, чтобы не зависили от классов. поменяла имена _check_is_ranking_loss / _check_is_regressor_loss на _check_is_compatible_loss
* Добавила _to_subclass

**Comment by nikitxskv** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
CLA already signed

**Comment by arcadia-devtools** [comment]
@nikitxskv, internal review request created: [1747136](https://a.yandex-team.ru/review/1747136)

**Comment by nikitxskv** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
@nikitxskv, internal review request updated: [1747136](https://a.yandex-team.ru/review/1747136)

**Comment by dremovd** [comment]
Попробовал поставить последнюю версию пакета из pip (0.25.1), но в ней еще нет CatBoostRanker. Так и должно быть?
